W 12190419 dwformats docchk v4 duke (#225)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-formats-flatfile.adoc

// CONFIG PROPS GET APPENDED BELOW /////////////////////////////////

> 

Author: TODO: compare with the include and remember to update with new property in Jan v2.4 patch update.

[[properties]]
== Configuration Properties

DataWeave supports the following configuration properties for this format.

[[reader_properties]]
=== Reader Properties

This format accepts properties that provide instructions for reading input data.

> 

Author: Note: To see if there have been updates since 2.4.0, checked the diff on the support/2.4.0 and support/2.5.0 build of the format docs. There was only one difference, which suggests there was no significant modification.  

```
< |`bufferSize`  |`Number`|`8192`|Size of the buffer writer. The value must be greater than 8.
< 
---
> |`bufferSize`  |`Number`|`8
```192`|Size of the buffer writer. The value needs to be bigger than 8.

Author: added to include

[cols="1,1,1,3a", options="header"]
|===
|Parameter |Type |Default|Description

> |`allowLenientWithBinaryNotEndElement`  |`Boolean`|`false`|When the schema contains elements of type Binary or Packed, the `lenient` option does not allow short records, regardless of the last element's

Author: Reader or Writer property? 
Is `allowLenientWithBinaryNotEndElement` a reader or writer property. *The 2.4 doc shows it as a writer property.* See https://docs.mulesoft.com/dataweave/2.4/dataweave-formats-fixedwidth#properties

Author: added question to include

Author: Tomo confirmed that it is Reader property. 
Need to fix in the docs (all versions where it's listed).

Author: added to v2.4

Author: done in 2.5 of course

format type. When you set this property to `true`, the validation applies only to cases in which the record ends with type Binary or Packed.

Valid values are `true` or `false`.

> |`enforceRequires`  |`Boolean`|`false`|Produces an error when set to `true` if a required value is missing.

Author: `| `enforceRequires` | `Boolean` | `false` | Error if required value missing.`

Author: new version is better

Valid values are `true` or `false`.

> |`missingValues`  |`String`|`null`|Fill character used to represent missing values. To activate a non-default setting, set the `useMissCharAsDefaultForFill` property to `true`, and use one of the following values to `missingValues`:

Author: `| `missingValues` | `String` | `nulls` for copybook schema, `spaces` otherwise | Fill character used to represent missing values. To activate a non-default setting, set the `useMissCharAsDefaultForFill` property to `true`, and use one of the following values to `missingValues`:

* `none`: Treat all data as actual values
* `spaces`: Interpret a field consisting of only spaces as a missing value
* `zeroes`: Interpret numeric fields consisting of only '0' characters and
character fields consisting of only spaces as missing values
* `nulls`: Interpret a field consisting only of 0 bytes as a missing value`

Author: the new descriptions are better

> |`missingValues`  |`String`|`null`|Fill character used to represent missing values. To activate a non-default setting, set the `useMissCharAsDefaultForFill` property to `true`, and use one of the following values to `missingValues`:

Author: ```suggestion
|`missingValues`  |`String`| `nulls` for COPYBOOK, `spaces` otherwise |Fill character used to represent missing values. To activate a non-default setting, set the `useMissCharAsDefaultForFill` property to `true`, and use one of the following values to `missingValues`:
```

Author: need this fix in the include

Author: this is already in the include

* `none` (for the reader) or `NONE` (for the writer): Treats all data as values.
* `spaces` (for the reader) or `SPACES` (for the writer): Interprets a field consisting of only spaces as a missing value. Default for flat file and fixed-width formats.
* `zeroes` (for the reader) or `ZEROES` (for the writer): Interprets numeric fields consisting of only `0` characters _and_ character fields consisting of only spaces as missing values.
* `nulls` (for the reader) or `NULLS` (for the writer): Interprets a field consisting only of `0` bytes as a missing value. Default for COBOL copybook schema.

> Valid values are `none` or `spaces` or `zeroes` or `nulls`.

Author: ```suggestion
```

Already listed.

Author: removed from include

> |`notTruncateDependingOnSubjectNotPresent`  |`Boolean`|`false`|Fills the entire group when the DEPENDING ON subject is not present.

Author: `notTruncateDependingOnSubjectNotPresent` not in 2.4. Is this new?

Author: It was introduced in a Mule 4.3.0/DW 2.3.0 patch update. See https://docs.mulesoft.com/release-notes/mule-runtime/mule-4.3.0-release-notes#august-2021

Patch release version: 4.3.0-20210622

Author: TODO: add to include

Author: added to 2.4

Author: todo: add to 2.3

Valid values are `true` or `false`.

> |`recordParsing`  |`String`|`'strict'`|Specifies the expected type of separation between lines or records:

Author: `| `recordParsing` | `String` | `strict` |

Expected separation between lines/records:

* `strict`: line break expected at exact end of each record
* `lenient`: line break used but records may be shorter or longer than schema specifies
* `noTerminator`: records follow one another with no separation
* `singleRecord`: entire input is a single record

Note that schemas with type `Binary` or `Packed` don't allow for line break
detection, so setting `recordParsing` to `lenient` only allows long records
to be handled, not short ones. These schemas also currently only work with
certain single-byte character encodings
(so not with UTF-8 or any multibyte format).`

Author: The wording could be improved a bit in the source if time permits, but not a priority.

* `strict`: Line break is expected at exact end of each record. `strict` is the default.
* `lenient`: Line break is used, but records can be shorter or longer than the schema specifies.
* `noTerminator`: Records follow one another with no separation.
* `singleRecord`: The entire input is a single record.

> Note that schemas with type `Binary` or `Packed` don't allow for line break detection, so setting `recordParsing` to `lenient` only allows long records to be handled, not short ones. These schemas also currently only work with certain single-byte character encodings (so not with UTF-8 or any multibyte format).

Author: The wording could be improved a bit in the source if time permits, but not a priority.

Valid values are `strict` or `lenient` or `noTerminator` or `singleRecord`.

> |`retainEmptyStringFieldsOnParsing`  |`Boolean`|`false`|Allow parsing behavior to keep missing string value fields with a default value in the output map

Author: `retainEmptyStringFieldsOnParsing` not in 2.4. Is this new? Check description and which of the 3 formats it applies to.

Author: Introduced in https://docs.mulesoft.com/release-notes/mule-runtime/mule-4.3.0-release-notes#january-2023

W-12036336

There's a separate PR with this for the v2.4 (and v2.3 doc): https://github.com/mulesoft/docs-dataweave/pull/218

Author: TODO: add to include

Author: handling the rest of this via #218

Valid values are `true` or `false`.

> |`schemaPath` (Required) |`String`|`null`|Path to the schema definition. Specifies the location in your local disk of the schema file that parses your input.

Author: `| `schemaPath`  (Required) | `String` | `null` | Schema definition. Location in your local disk of the schema file used to parse your input.`

> |`segmentIdent`  |`String`|`null`|Segment identifier in the schema for fixed-width or COBOL copybook schemas. Required when parsing a single segment or record definition if the schema includes multiple segment definitions.

Author: `| `segmentIdent` | `String` | `null` | Segment identifier in the schema for fixed width or copybook schemas (only needed when parsing a single segment/record definition and if the schema includes multiple segment definitions).`

Author: better in new doc

> |`structureIdent`  |`String`|`null`|Structure identifier in the schema for flat file schemas. Required when parsing a structure definition if the schema includes multiple structure definitions.

Author: `| `structureIdent` | `String` | `null` | Structure identifier in schema for flatfile schemas (only needed when parsing a structure definition, and if the schema includes multiple structure definitions)`

Author: better in output

> |`truncateDependingOn`  |`Boolean`|`false`|For COBOL copybook, truncates DEPENDING ON values to the length used.

Author: `| `truncateDependingOn` | `Boolean` | `false` | Truncate COBOL copybook DEPENDING ON values to the length used. Valid values are `true` or `false`.`

Author: better in output

Valid values are `true` or `false`.

> |`useMissCharAsDefaultForFill`  |`Boolean`|`false`|By default, the flat file reader and writer use spaces for missing characters and ignore the setting of the `missingValues` property. When you set this property to `true`, DataWeave honors the setting of the `missingValues property`. Introduced in DataWeave 4.3 (2.3.0-20210823) for the September 2021 release of Mule 4.3.0-20210823.

Author: `| `useMissCharAsDefaultForFill`|`Boolean`| `false` |By default, the flat file reader and writer use spaces for missing characters and ignore the setting of the `missingValues` property. When you set this property to `true`, the reader honors the setting of the `missingValues` property.  _Introduced in DataWeave 4.3 (2.3.0-20210823) for the September 2021 release of Mule 4.3.0-20210823._ Valid values are `true` or `false`.`

Valid values are `true` or `false`.

> |`zonedDecimalStrict`  |`Boolean`|`false`|For COBOL copybook, uses the 'strict' ASCII form of sign encoding for zoned decimal values.

Author: `| `zonedDecimalStrict` | `Boolean` | `false` | Use the `strict` ASCII form of
sign encoding for COBOL copybook zoned decimal values.
Valid values are `true` or `false`.`

Author: looks fine in output

Valid values are `true` or `false`.
|===

[[writer_properties]]
=== Writer Properties

This format accepts properties that provide instructions for writing output data.

[cols="1,1,1,3a", options="header"]
|===
|Parameter |Type |Default|Description

> |`bufferSize`  |`Number`|`8192`|Size of the buffer writer. The value must be greater than 8.

Author: This says, "The value must be greater than 8." but the default is 8192. What unit is this in? 

The 2.4 docs do not have that sentence.

Author: asked for more info on dw channel

Author: done

> |`bufferSize`  |`Number`|`8192`|Size of the buffer writer. The value must be greater than 8.

Author: ```suggestion
|`bufferSize`  |`Number`|`8192`|Size of the buffer writer, in bytes. The value must be greater than `8`.
```

> |`deferred`  |`Boolean`|`false`|Generates the output as a data stream when set to `true`, and defers the script's execution until the generated content is consumed.

Author: 2.4 doc:

`| `deferred` | `Boolean` | `false` | When set to `true`, DataWeave generates the output as a data stream, and the script's execution is deferred until it is consumed.
  Valid values are `true` or `false`.`

Author: looks fine

Valid values are `true` or `false`.

> |`encoding`  |`String`|`null`|The encoding to use for the output, such as UTF-8.

Author: 2.4 doc
`| `encoding` | `String` | `null` | Encoding to be used by this writer,
such as `UTF-8`.`

Author: looks good

> |`enforceRequires`  |`Boolean`|`false`|Produces an error when set to `true` if a required value is missing.

Author: `| `enforceRequires` | `Boolean` | `false` | Error if a required value is missing.
Valid values are `true` or `false`.`

Author: looks fine

Valid values are `true` or `false`.

> |`missingValues`  |`String`|`null`|Fill character used to represent missing values. To activate a non-default setting, set the `useMissCharAsDefaultForFill` property to `true`, and use one of the following values to `missingValues`:

Author: `| `missingValues` | `String` | `NULLS` for copybook schema, `SPACES` otherwise | Fill character used to represent missing values. To activate a non-default setting, use `useMissCharAsDefaultForFill`, and use one of the following values to `missingValues`:`

Author: output looks better

> * `none` (for the reader) or `NONE` (for the writer): Treats all data as values.

Author: ```suggestion
* `NONE`: Write nothing for missing values.
```

Author: TBD

> * `spaces` (for the reader) or `SPACES` (for the writer): Interprets a field consisting of only spaces as a missing value. Default for flat file and fixed-width formats.

Author: `* `SPACES`: Fill the field with spaces.`

Author: TBD for reader and writer

> * `zeroes` (for the reader) or `ZEROES` (for the writer): Interprets numeric fields consisting of only `0` characters _and_ character fields consisting of only spaces as missing values.

Author: `* `ZEROES`: Fill numeric fields with `0` characters and character fields with space characters.`

Author: TBD for reader and writer

> * `nulls` (for the reader) or `NULLS` (for the writer): Interprets a field consisting only of `0` bytes as a missing value. Default for COBOL copybook schema.

Author: `* `NULLS`: Fill the field with 0 bytes.`

Author: TBD for reader and writer

Valid values are `none` or `spaces` or `zeroes` or `nulls`.

> |`notTruncateDependingOnSubjectNotPresent`  |`Boolean`|`false`|Fills the entire group when the DEPENDING ON subject is not present.

Author: Seems to be new in 2.5

Author: It was introduced in a Mule 4.3.0/DW 2.3.0 patch update. See https://docs.mulesoft.com/release-notes/mule-runtime/mule-4.3.0-release-notes#august-2021

Patch release version: 4.3.0-20210622

TODO: add to the include

Author: explainer: 4.4.0 was released Oct 2021, and the patch update to 4.3.0 with this property was in Aug 2021, which explains why the property appears in the 4.4.0 and 4.5.0 but not in the 4.4.0 RNs.

Author: TODO: add the property to the 4.3.0 doc, explaining which patch it was added in

Author: Added to v4.3

Valid values are `true` or `false`.

> |`recordTerminator`  |`String`|`null`|Line break for a record separator.

Author: In 2.4:
`| `recordTerminator` | `String` | System property `line.separator` | Record separator line break. Valid values:

* `lf`
* `cr`
* `crlf`
* `none`

Note that in Mule versions 4.0.4 and later, this is only used as a separator
when there are multiple records. Values translate directly to character codes
(`none` leaves no termination on each record).`

Author: TBD: looks like the default needed manual fix

Author: TODO: verify the default and note for 2.5.0

Author: TODO: how can the default be null when that's not listed as a valid value? confusing

Author: tweak to v2.4, adding period

Author: added to master

> |`recordTerminator`  |`String`|`null`|Line break for a record separator.

Author: ```suggestion
|`recordTerminator`  |`String`|`null`|Line break for a record separator. Valid values: `lf`, `cr`, `crlf`, `none`. Note that in Mule versions 4.0.4 and later, this is only used as a separator when there are multiple records. Values translate directly to character codes (`none` leaves no termination on each record).
```

Author: added to master

> Valid values are `lf` or `cr` or `crlf` or `none`.

Author: ```suggestion
```

> |`schemaPath` (Required) |`String`|`null`|Path to the schema definition. Specifies the location in your local disk of the schema file that parses your input.

Author: `| `schemaPath` (Required) | `String` | `null` | Schema definition. Path where the schema file
to be used is located.`

Author: output looks better

> |`segmentIdent`  |`String`|`null`|Segment identifier in the schema for fixed-width or COBOL copybook schemas. Required when parsing a single segment or record definition if the schema includes multiple segment definitions.

Author: `| `segmentIdent` | `String` | `null` | Segment identifier in the schema for fixed width or copybook schemas (only needed when writing a single segment/record definition, and if the schema includes multiple segment definitions).`

Author: looks fine

> |`structureIdent`  |`String`|`null`|Structure identifier in the schema for flat file schemas. Required when parsing a structure definition if the schema includes multiple structure definitions.

Author: `| `structureIdent` | `String` | `null` | Structure identifier in schema for flatfile schemas (only needed when writing a structure definition and if the schema includes multiple structure definitions)| `trimValues` | `Boolean` | `false` | Trim string values longer than the field length by truncating trailing characters. Valid values are `true` or `false`.`

Author: looks fine

> |`trimValues`  |`Boolean`|`false`|Trim values that are longer than the width of a field.

Author: `| `trimValues` | `Boolean` | `false` | Indicates whether trim values are longer than the field width. Valid Options are `true` or `false`.`

Author: output is better

Valid values are `true` or `false`.

> |`truncateDependingOn`  |`Boolean`|`false`|For COBOL copybook, truncates DEPENDING ON values to the length used.

Author: `| `truncateDependingOn` | `Boolean` | `false` | Truncate COBOL copybook DEPENDING ON values to the length used. Valid values are `true` or `false`.`

Author: looks fine

Valid values are `true` or `false`.

> |`useMissCharAsDefaultForFill`  |`Boolean`|`false`|By default, the flat file reader and writer use spaces for missing characters and ignore the setting of the `missingValues` property. When you set this property to `true`, DataWeave honors the setting of the `missingValues property`. Introduced in DataWeave 4.3 (2.3.0-20210823) for the September 2021 release of Mule 4.3.0-20210823.

Author: `| `useMissCharAsDefaultForFill`|`Boolean`| `false` |By default, the flat file reader and writer use spaces for missing characters and ignore the setting of the `missingValues` property. When you set this property to `true`, the writer honors the setting of the `missingValues` property.  _Introduced in DataWeave 4.3 (2.3.0-20210823) for the September 2021 release of Mule 4.3.0-20210823._
`

Author: looks good

Valid values are `true` or `false`.

> |`zonedDecimalStrict`  |`Boolean`|`false`|For COBOL copybook, uses the 'strict' ASCII form of sign encoding for zoned decimal values.

Author: `| `zonedDecimalStrict` | `Boolean` | `false` | Use the `strict` ASCII form of
sign encoding for COBOL copybook zoned decimal values. Valid values are `true` or
`false`.`

Author: looks fine
